### PR TITLE
[InputGen] overview

### PR DESCRIPTION
**Pull request by manuelcandales**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116138
* #116137
* #116136
* #116135
* #115698
* #115697
* #115696
* #115695
* #115694
* #115693
* #115692
* #115691
* #115605
* #115604
* #115603
* #115602
* __->__ #115601

InputGen's objective is to generate inputs to torch ops, given certain constraints, which specify when some input is valid for the op and when it is not.

Differential Revision: [D51816907](https://our.internmc.facebook.com/intern/diff/D51816907/)

### CODE REVIEW

File: torch/testing/_internal/inputgen/overview.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+### MetaArg
+
+When generating an argument, we first generate *everything except the tensor data* (if we are dealing with a tensor or a list of tensors). That "everything except the tensor data" is what we call the meta argument. For an int, or a boolean or a Scalar, the meta argument contains all of the same information that the argument contains, but organized a bit differently.
+
+A meta argument contains the *structure* of the argument (think about rank, and sizes in the case of a tensor, or the length of a list of integers). It also contains the dtype of the argument (if talking about tensors or scalars). It also contains the value of the argument (if talking about anything except a tensor or list of tensors)
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torch/testing/_internal/inputgen/overview.md]: Updated that section

File: torch/testing/_internal/inputgen/overview.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+### Attributes
+
+Arguments have attributes. A tensor has dtype, rank, sizes and values. A list of integers has length and values. A scalar has dtype and value. A list of tensors has length, dtypes, ranks, sizes and values. We have consolidated argument qualities behind 6 attributes: Optional, Dtype, Length, Rank, Size, Value.
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torch/testing/_internal/inputgen/overview.md]: Updated that section. Should be more clear now.

File: torch/testing/_internal/inputgen/overview.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+InputGen's objective is to generate inputs to torch ops, given certain constraints, which specify when some input is valid for the op and when it is not. InputGen consists of such set of specifications for each argument of an op, and an input generation engine.
+
+## Arguments
+
+Before giving an overview of the specifications and the generation engine, let's first examine the argument themselves.
+
+### ArgType
+
+To being describing an argument, we first need to declare its type. An argument might be a Tensor, an Optional Tensor, a List of Tensors, a Scalar. It could also be just a boolean, or a float, or an int, or a list of int, etc. However, there a "different types" of int. For example, an int argument might be a dimension, or could be an index, or could be how much pad to add on each side. There are different constraints imposed on the "int" depending on the case. Therefore, we decided to introduce the ArgType abstraction. Different ArgTypes are: Tensor, Scalar, Dim, DimList, Index, Length, etc.
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torch/testing/_internal/inputgen/overview.md]: I am not sure about this one. I don't want to have two places that need to be updated, if a new ArgType needs to be added in the future.

[the remark this answers is not in the document]
Comment (manuelcandales) [reply torch/testing/_internal/inputgen/overview.md]: By the way, I did incorporate your suggestions above in the last commit.

**Review by lezcano — COMMENTED** [review]
This is a rather large piece of engineering.

I still have similar concerns to the ones we discussed during the conference
1. This is a brute force approach towards testing ops vs. the manual one we use. This is surely going to find many little bugs, at the cost of running *many more* samples. What is the plan for making this scale? Where is this going to run? Running this on everything that runs OpInfos (functorch / {grad, fwd AD}  x {grad fwd AD} / amp / all backends / all dtypes / ...) does not seem like a feasible approach
2. If the target of this project is to remove the 19k lines of definitions in OpInfos, then it'd be good to see a number of PRs in the stack that do this. A warning is in order: I would expect this task to be **very** time consuming as there are many tests that are not particularly numerically stable
3. This generates valid samples for the forward but not for the backward.

In summary, these PRs effectively implement a fuzzer for PyTorch. Historically, the issues coming from reserach groups that have implemented fuzzers have not been of particularly high quality. How is `InputGen`  different to those previous approaches and how does it plan to tackle this?

I also wonder whether we would want to land this into core or its own separate repo. I do see value on having a way to get a per-op sampler, but I wonder whether core is the right place for it given the issues discussed above.

**Comment by manuelcandales** [comment]
Hi, thank you for taking a look at this. I am happy to read your opinion on this. Although, in the conference you seemed more exited about this project than you seem to be now. Ok, let me address your comments.

> 1. This is a brute force approach towards testing ops vs. the manual one we use.

This is **not** a brute force approach. It is somewhat subtle, and probably this deserves a detailed explanation in the overview. This is an algorithmic approach, but is not done through brute force.
The database of specifications, provides a complete specification of the validity space of the arguments. So, this approach is pretty spear-headed. The result is that we generate values that we know are supposed to be valid and we know are edge cases. We also generate values that we know are supposed to be invalid and we know are edge cases. It also generates values that aren't edge cases. But the point is that it can differentiate between them.

>This is surely going to find many little bugs, at the cost of running _many more_ samples.

We are talking about less than 100 samples for smaller ops, maybe a few hundred samples for larger ops. We are not talking of a combinatorial explosion of samples. Again, this is not brute forcing.

> What is the plan for making this scale?

A few hundred samples per op can't scale? If it can't, then we might have to take random samples from those few hundred. But I think a few hundred should be fine.

> Where is this going to run? Running this on everything that runs OpInfos (functorch / {grad, fwd AD}  x {grad fwd AD} / amp / all backends / all dtypes / ...) does not seem like a feasible approach

This is not intended to replace OpInfos. Maybe we can eventually run it on everything that runs OpInfos, maybe not. I have not looked into the backward side of things. For now, this is only focused on the forward.
It is intended to find bugs on existing forward implementations, and to aid people writing new forward implementations of existing ops. With this tool, you can test your implementations of Core ATen ops, and make sure they are correct.

> 2. If the target of this project is to remove the 19k lines of definitions in OpInfos, then it'd be good to see a number of PRs in the stack that do this. A warning is in order: I would expect this task to be **very** time consuming as there are many tests that are not particularly numerically stable

No, this is not the target. If eventually, after this project has matured, we find that it can replace all of the use cases of OpInfos, then so be it. For now, this is not the intent.

> 3. This generates valid samples for the forward but not for the backward.

This is true. Only focused on the forward at the moment. But again, I learned to multiply before I learned to divide.

> In summary, these PRs effectively implement a fuzzer for PyTorch. Historically, the issues coming from reserach groups that have implemented fuzzers have not been of particularly high quality. How is `InputGen` different to those previous approaches and how does it plan to tackle this?

I would not call InputGen a fuzzer. But I don't want to get down to semantics of what is a fuzzer and what is not. If it gets classified as a fuzzer, then so be it, but at the very least, it is a very smart fuzzer. Because it knows everything there is to know about the op. It is not blindly generating values, and see what works.

One very important note, this stack is not yet complete. I have not submitted the diff with the database of specifications. That's were the value lies. You are asking how is `InputGen` different to previous approaches. I am not familiar with those fuzzers you are referring to, so, I can't draw comparisons. Did those previous approaches employed complete and detailed specifications for op arguments?

This tool gives you great control when generating inputs. As a matter of fact, that is the one thing this tool is designed to do. Maybe you need those inputs for testing, maybe you need them for something else
You can say: hey, I want valid inputs for convolution, such that the weight tensor has rank 4, its 1st dimension is even, and the bias has dtype torch.int16. And if that set of constraints is impossible to achieve, it will tell you, with mathematical certainly, hey, those constraints have an empty set of solutions. And if it is solvable, it can produce sample inputs that match those constraints.

> I also wonder whether we would want to land this into core or its own separate repo. I do see value on having a way to get a per-op sampler, but I wonder whether core is the right place for it given the issues discussed above.

I am interested in hearing again your take on this, after you read my answers above.

**Comment by kimishpatel** [comment]
I havent read through @manuelcandales's response to @lezcano but if it seems like something that can be better discussed over a meeting, please do schedule one. I do love having this discussion in OSS but also want to make sure we can resolve concerns quickly.
